File: desktop/macos/Sources/Localyze/Artifacts/VizParser.swift
import Foundation

enum VizKind: String { case chart, table, map, run, code, form, image, pdf }

struct VizBlock {
    let kind: VizKind
    let attrs: [String: String]
    let inner: String?
}

/// Streaming-friendly parser for <viz ...> blocks.
/// Spec: ../../shared/viz-schema.md
enum VizParser {
    static func parse(_ text: String) -> [VizBlock] {
        var out: [VizBlock] = []
        let pattern = #"<viz\s+([^>]+?)(?:/>|>(.*?)</viz>)"#
        guard let rx = try? NSRegularExpression(pattern: pattern, options: [.dotMatchesLineSeparators]) else { return [] }
        let ns = text as NSString
        rx.enumerateMatches(in: text, range: NSRange(location: 0, length: ns.length)) { m, _, _ in
            guard let m = m else { return }
            let attrsStr = ns.substring(with: m.range(at: 1))
            let inner = m.numberOfRanges > 2 && m.range(at: 2).location != NSNotFound
                ? ns.substring(with: m.range(at: 2)) : nil
            let attrs = parseAttrs(attrsStr)
            guard let typeStr = attrs["type"], let kind = VizKind(rawValue: typeStr.lowercased()) else { return }
            out.append(VizBlock(kind: kind, attrs: attrs, inner: inner))
        }
        return out
    }

    private static func parseAttrs(_ s: String) -> [String: String] {
        var d: [String: String] = [:]
        let pattern = #"(\w+)\s*=\s*(?:"([^"]*)"|'([^']*)')"#
        guard let rx = try? NSRegularExpression(pattern: pattern) else { return d }
        let ns = s as NSString
        rx.enumerateMatches(in: s, range: NSRange(location: 0, length: ns.length)) { m, _, _ in
            guard let m = m else { return }
            let k = ns.substring(with: m.range(at: 1))
            let v = m.range(at: 2).location != NSNotFound
                ? ns.substring(with: m.range(at: 2))
                : ns.substring(with: m.range(at: 3))
            d[k] = v
        }
        return d
    }
}

File: desktop/windows/Localyze/Artifacts/VizParser.cs
using System.Text.RegularExpressions;

namespace Localyze.Artifacts;

public enum VizKind { Chart, Table, Map, Run, Code, Form, Image, Pdf }

public sealed record VizBlock(VizKind Kind, IReadOnlyDictionary<string, string> Attrs, string? Inner);

/// <summary>
/// Streaming parser for &lt;viz type="..." .../&gt; blocks. The parser is intentionally
/// permissive: it tolerates partial JSON in attributes so charts/tables can render
/// incrementally as the model writes data. Spec: desktop/shared/viz-schema.md
/// </summary>
public static class VizParser
{
    private static readonly Regex BlockRx = new(
        @"<viz\s+(?<attrs>[^>]+?)(?:/>|>(?<inner>.*?)</viz>)",
        RegexOptions.Compiled | RegexOptions.Singleline);

    public static IEnumerable<VizBlock> Parse(string text)
    {
        foreach (Match m in BlockRx.Matches(text))
        {
            var attrs = ParseAttrs(m.Groups["attrs"].Value);
            if (!attrs.TryGetValue("type", out var type)) continue;
            if (!System.Enum.TryParse<VizKind>(type, true, out var kind)) continue;
            var inner = m.Groups["inner"].Success ? m.Groups["inner"].Value : null;
            yield return new VizBlock(kind, attrs, inner);
        }
    }

    private static Dictionary<string, string> ParseAttrs(string s)
    {
        var d = new Dictionary<string, string>(System.StringComparer.OrdinalIgnoreCase);
        foreach (Match m in Regex.Matches(s, "(?<k>\\w+)\\s*=\\s*(?:\"(?<v1>[^\"]*)\"|'(?<v2>[^']*)')"))
            d[m.Groups["k"].Value] = m.Groups["v1"].Success ? m.Groups["v1"].Value : m.Groups["v2"].Value;
        return d;
    }
}
